Create imu launch pkg, setup and add Dockerfile

GreatAlexander opened the pull request.

GreatAlexander committed:
Create imu launch pkg, setup and add Dockerfile

GreatAlexander committed:
Update Dockerfile with release compilation & more

GreatAlexander committed:
Upgrade Dockerfile w/ stages & convenient scripts

hect95 commented on av_imu_launch/config/av_gx5_15.yml:
```diff
@@ -0,0 +1,15 @@
+    port : '/dev/imu-front'
+    baudrate : 230400
+
+    # This will cause the node to convert any NED measurements to ENU
```
Codespell hook fails due to "NED" word caught as typo:

`v_imu_launch/config/av_gx5_15.yml:7: NED ==> NEED, END, NOD`

Maybe we put explicitly the acronym "North-East-Down" ? Alternatively we can just ignore it in the .pre-commit-config.yaml

GreatAlexander replied:
Indeed, the pre-commit (Which I got from [elsewhere](https://github.com/ros-controls/ros2_control/blob/master/.pre-commit-config.yaml)) is a little over-zealous right now.
I would add any original source code (e.g. from a fork) to be ignored by the .pre-commit (We're not trying to fix upstream source code, and we don't want to change it in our fork to avoid any conflicts). I forked a bunch of repos at the start for testing which may no longer need the source code, so that also needs doing.

For this particular pre-commit typo, I'll just reword the comment as you suggest. I would like to run pre-commit for all our work when applicable with few exceptions (e.g. I added [*.dae to be ignored](https://github.com/ipab-rad/car_description/pull/1/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R131) in the car-description repo).

GreatAlexander replied:
Changed my mind :sweat_smile: I'm adding it to the list of codespell ignored words since we "could" (Given it's an IMU launch repo) use the acronym NED with reasonable context and it will be picked up _again_.

hect95 commented on Dockerfile:
```diff
@@ -0,0 +1,36 @@
+        # Ping network tools
+        inetutils-ping \
+        # Install Lord IMU driver pkg
+        ros-$ROS_DISTRO-microstrain-inertial-driver \
```
Given that this is a sensor driver, maybe we should explicitly state the pkg version just to avoid potential problems?

The docker gets build every time the `run.sh`, `dev.sh` and `build.sh` scripts are used. A potential new driver update will pass inadvertently specially after the `apt update`.

I run `apt list -a ros-humble-microstrain-inertial-driver` inside the docker and it seems the version you probably tested was `3.2.1-1jammy.20240304.151451`

GreatAlexander replied:
I flip-flopped about this a few times in the past, with pros/cons for pinning a specific version (In the past, we tended to pin specific versions _only_ iff no newer would do [since you assume no major interface breaks unless it's a major version change]). In fact recently (I think it was the spinnaker camera drivers) I was keen to get the newest driver asap because they were actively fixing an issue with the driver.

However, probably a better practice for big complex code-bases with multiple components [such as an AV stack] is to pin everything just in case, so I think this is how we should do it from now on. I'll add this pinned version (And we should do so from now on in all docker containers). I presume we should only pin hard dependencies like drivers & libraries, we can probably get away without pinning e.g. `nano` for the dev images.

hect95 replied:
Agreed!

hect95 commented on av_imu_launch/package.xml:
```diff
@@ -0,0 +1,17 @@
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>av_imu_launch</name>
+  <version>0.0.0</version>
```
Shall we define the package version as `1.0.0`? I just thought it would be a good idea for our reference in the future.

GreatAlexander replied:
Indeed, the very first time I was setting all of this up I just wanted to go with 0s until we made the first "release", which getting all of these initial PRs merged may qualify as.
Back in my day we had these version numbers automatically generated and populated given our changelog updates.
It was convenient, but I guess we don't have to go to such lengths of automation (yet).
There are multiple ways we could go about it, I don't think we will be publicising our work often, but it will be very convenient indeed to generate docker images which versions correspond to the pkg versions within them.

Some relevant reading:
https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes
https://medium.com/fiverr-engineering/major-minor-patch-a5298e2e1798
https://github.com/audrow/update-ros-pkg-versions

hect95 replied:
I was thinking about the same. It wouldn't hurt to just semi-automate something so we ensure proper versioning for the good practice

hect95 commented on README.md:
```diff
@@ -1,2 +1,6 @@
+
+## Info
+
+Lord GX5-3DM-15 IMU
```
I think worth to specify here the IMU driver dep and version

GreatAlexander replied:
Roger roger

hect95 commented on av_imu_launch/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+  DESTINATION share/${PROJECT_NAME}
+)
+
+if(BUILD_TESTING)
```
I do not think we are going to define any c++ code in this pkg, we can simplify the CMakeLists.txt

GreatAlexander replied:
Soz I'm a ROS2 newbie and a CMake noob and copy-pasted the wrong file, simplified now.

hect95 commented on README.md:
```diff
@@ -1,2 +1,6 @@
+
+## Info
+
+Lord GX5-3DM-15 IMU
```
Would be good if we can explain the docker scripts and the use of them (we can copy and paste this description across other AV repos).

As a context: For me when I see `build.sh` and `run.sh` along with a dockerfile, I immediately think that `build.sh` is for literally only building the docker and `run.sh` is to run the built image (and potentially bash CMD accepted as args). Instead these scripts always build the docker image up to a certain target - either `build` and `dev`.

hect95 replied:
I have added an #Usage section

hect95 commented on run.sh:
```diff
@@ -0,0 +1,14 @@
+# Build docker full image and run ROS code interactively for debugging        #
+###############################################################################
+
+# Build docker image only up to base stage
+DOCKER_BUILDKIT=1 docker build \
+-t imu_humble \
+-f Dockerfile --target build .
+
+# Run docker image without volumes
+docker run -it --rm --net host \
+-v /dev/shm:/dev/shm \
+imu_humble
```
I think this is very minor but, potentially, `run.sh` can accept an arg that specifies `dev` or `runtime` docker run.

Also, do we want to build the docker image every time we run the container? I think we can save time if we do not, specially if the amount of docker containers that defined our AV stack increases.

Let me know your comments on this one.

hect95 replied:
`runtime.sh` and `dev.sh` docker convenience scripts have been implemented.

hect95 reviewed (commented):
Leaving few minor comments :)

GreatAlexander committed:
Multiple minor fixes from PR comments:

 - Add NED to codespell ignore names
 - Pin IMU driver version in Dockerfile dependency apt install
 - Update README.md with more information
 - Clean CMakeLists.txt given we compile no C++ code

hect95 committed:
Updating Docker scripts and dockerfile to align with Docker multi-stage building convention

- Dockerfile now implements base, prebuilt, dev and runtime stage names convention
- run.sh wwill run the runtime container with bash if an argument is defined. `build.sh` scritp is removed due to ambiguity
- Updating README to incude "usage" section

hect95 committed:
Renaming run.sh into runtime.sh to keep consistency with docker multi-stage build names; Updating README

hect95 commented:
This PR can be merged after testing on vehicle :)

hect95 commented on av_imu_launch/package.xml:
```diff
@@ -0,0 +1,17 @@
+  <name>av_imu_launch</name>
+  <version>0.0.0</version>
+  <description>AV velodyne config and launch</description>
+  <maintainer email="alex.bordallo@ed.ac.uk">Alejandro Bordallo</maintainer>
```
Could you please add me as maintainer? :)

GreatAlexander committed:
Add Hector Cruz as maintainer

hect95 reviewed (dismissed):
Good to go!

hect95 committed:
Adding privileged docker run

hect95 committed:
Mounting usb /dev dir into container

hect95 reviewed (approved):
Tested on vehicle so good to go!
